src/solver/algorithms/norvig_solver.py

```diff
@@ -0,0 +1,145 @@
+# norvig_solver.py
+# author: Josue Mendoza
+# date: 4-12-2015
+
+import collections
+from singleton import Singleton
+
+
+class NorvigSolver(object):
+    """NorvigSolver instance objects have attributes containing values
+    required for the norvigs algorithm to solve a sudoku game.
+    Can there only be one instance of this object.
+    """
+    __metaclass__ = Singleton
+
+    def __init__(self):
+        self.digits = '123456789'
+        self.rows = 'ABCDEFGHI'
+        self.cols = self.digits
+        self.squares = self.cross(self.rows, self.cols)
+        self.unit_list = ([self.cross(self.rows, col) for col in self.cols] +
+                        [self.cross(row, self.cols) for row in self.rows] +
+                        [self.cross(row_square, col_square) for row_square in ('ABC','DEF','GHI')
+                        for col_square in ('123','456','789')])
+        self.units = dict((square, [unit for unit in self.unit_list if square in unit])
+                        for square in self.squares)
+        self.peers = dict((square, set(sum(self.units[square],[]))-set([square]))
+                        for square in self.squares)
+
+    def cross(self, list_A, list_B):
+        """Cross product of elements in A and elements in B.
+        Keyword arguments:
+        list_A -- list with elements to be combined
+        list_B -- list with elements to be combined
+        """
+        return [a+b for a in list_A for b in list_B]
+
+    def parse_grid(self, grid):
+        """Converts grid to a dictionary of possible values, {square: digits}, or
+        return False if a contradiction is detected.
+        Keyword arguments:
+        grid -- it is a list containing the grid values, it will have a zero ('0') for
+        non-set values (i.e. ['0', '0', '3', '0', '2', '0', '6', '0', '0', '9',...)
+        """
+        ## To start, every square can be any digit; then assign values from the grid.
+        values = dict((square, self.digits) for square in self.squares)
+        for square, digit in self.grid_values(grid).items():
+            if digit in self.digits and not self.assign(values, square, digit):
+                return False
+        return values
+
+    def grid_values(self, grid):
+        """Converts grid into a dict of {square: char} with '0' for empties.
+        Returns the dictionary generated (i.e. OrderedDict([('A1', '2'), ('A2', '0')...).
+        Keyword arguments:
+        grid -- it is a list containing the grid values, it will have a zero ('0') for
+        non-set values (i.e. ['0', '0', '3', '0', '2', '0', '6', '0', '0', '9',...)"""
+        assert len(grid) == 81
```

Comment:
this kind of validations it will be better to have them using custom exceptions that will record all these problems into the log